traits/has_traits.py

```diff
@@ -1425,9 +1428,12 @@ def __setstate__ ( self, state, trait_change_notify = True ):
         else:
             # Otherwise, apply the Traits 3.0 restore logic:
             self._init_trait_listeners()
+            inst_traits = state.pop('__instance_traits__', {})
             self.trait_set( trait_change_notify = trait_change_notify, **state )
             self._post_init_trait_listeners()
             self.traits_init()
+            for attr, trait in inst_traits.iteritems():
```

Comment:
@cfarrow I notice that here we're calling `add_trait` _after_ doing `traits_init()` and `_post_init_trait_listeners()`, but in the `clone_traits` block below, we're calling `add_trait` _before_ those calls. Do you know whether there's a reason for the difference?

Reply:
I don't know the reason. I'll look into the potential consequences. At the very least I'll make it consistent.

Reply:
According to the [doc string](https://github.com/enthought/traits/blob/master/traits/ctraits.c#L1440), `traits_init` should be called at the end of `__setstate__`.  In practice, I don't think it matters. It is a no-op by default. The only place I have found it defined in the traits source is in [HasUniqueStrings](https://github.com/enthought/traits/blob/master/traits/ustr_trait.py#L141). 

In any case, I moved it after all the `add_traits` to be consistent with the existing documentation and the behavior in `clone_traits`.

Reply:
Thanks, Chris. I also moved the `_post_init_trait_listeners` call, so that it's consistent with `clone_traits`.